pxtservices/iframeDriver.ts

```diff
@@ -0,0 +1,553 @@
+                else {
+                    pending.reject(resp.error || new Error("Unknown error: iFrame returned failure"));
+                }
+            }
+        }
+        else if (data.type === "pxthost") {
+            if (data.action === "editorcontentloaded") {
+                this.readyForMessages = true;
+                this.sendMessageCore(); // flush message queue.
+            }
+            else if (data.action === "workspacesync" || data.action === "workspacesave" || data.action === "workspacereset" || data.action === "workspaceloaded") {
+                this.handleWorkspaceSync(data as pxt.editor.EditorWorkspaceSyncRequest);
+            }
+
+            this.fireEvent(data.action, data);
+        }
+
+        this.fireEvent(MessageReceivedEvent, data);
+    }
+
+    protected fireEvent(event: string, data: any) {
+        const listeners = this.editorEventListeners[event];
+        if (!listeners) return;
+
+        for (const handler of listeners) {
+            try {
+                handler(data);
+            }
+            catch (e) {
+                console.error(e);
+            }
+        }
+    }
+
+    protected async handleWorkspaceSync(event: pxt.editor.EditorWorkspaceSyncRequest | pxt.editor.EditorWorkspaceSaveRequest) {
+        if (!this.host) return;
+
+        let error: any = undefined;
+        try {
+            if (event.action === "workspacesync") {
+                const status = await this.host.getWorkspaceProjects();
+                this.sendMessageCore({
+                    type: "pxthost",
+                    id: event.id,
+                    success: !!status,
+                    projects: status?.projects,
+                    editor: status?.editor,
+                    controllerId: status?.controllerId
+                } as pxt.editor.EditorWorkspaceSyncResponse);
+            }
+            else if (event.action === "workspacereset") {
+                await this.host.resetWorkspace();
+            }
+            else if (event.action === "workspacesave") {
+                await this.host.saveProject(event.project);
+            }
+            else if (event.action === "workspaceloaded") {
+                if (this.host.onWorkspaceLoaded) {
+                    await this.host.onWorkspaceLoaded();
+                }
+            }
+        }
+        catch (e) {
+            error = e;
+            console.error(e);
+        }
+        finally {
+            if (event.response) {
+                this.sendMessageCore({
+                    type: "pxthost",
+                    id: event.id,
+                    success: !error,
+                    error
+                } as pxt.editor.EditorMessageResponse);
+            }
+        }
+    }
+
+    protected sendRequest(message: any) {
+        return new Promise((resolve, reject) => {
+            message.response = true;
+            message.id = this.nextId++ + "";
+            this.pendingMessages[message.id] = {
```

Comment:
Noting a potential leak: All sent messages are wrapped in a promise and added to `pendingMessages`, which resolve once a reply is received. Is a reply received for every message sent?

Reply:
only requests are wrapped in a promise. all requests with `message.response` set to true should indeed receive a response from the editor. i guess we could time out requests, but i think it's a minor scenario and i worry it might affect some long running requests (like starting a tutorial on a low end chromebook)

i guess i could implement some sort of heartbeat mechanism where the iframe reports back what messages are still active, but that might be overkill... thoughts?

Reply:
Ok, that seems fine then. If we expect every message sent this way to receive a response, then promises won't leak except when something else is already broken. Adding a heartbeat seems like overkill :)